### DISCUSSION TITLE
Have plan to add support for File System Access API (OPFS)?

### DISCUSSION THREAD

**Discussion by hangxingliu** [duckdb/duckdb-wasm#1444] [Ideas]
[File System Access API](https://wicg.github.io/file-system-access/)  is now supported by most modern browsers. This signifies a more efficient means for frontend developers to access a large volume of persistent data compared to localStorage or IndexedDB.

This API offers a variety of beneficial methods such as  [createSyncAccessHandle](https://developer.mozilla.org/en-US/docs/Web/API/FileSystemFileHandle/createSyncAccessHandle) and [createWritable](https://developer.mozilla.org/en-US/docs/Web/API/FileSystemFileHandle/createWritable). These methods enable access to or writing parts of extensive persistent data from OPFS or the local filesystem (once the user has granted permission). Furthermore, these methods align well with many DuckDB use cases (e.g., exporting tables to files and data persistence requirements mentioned in recent discussions). Concurrently, SQLite3 WASM has incorporated this API, supplanting the obsolete WebSQL: https://developer.chrome.com/blog/sqlite-wasm-in-the-browser-backed-by-the-origin-private-file-system/

In summary, would you entertain the idea of supporting the File System Access API or devising a method to link DuckDB WASM with OPFS?
Thank you.

---

References:

- https://developer.chrome.com/blog/sqlite-wasm-in-the-browser-backed-by-the-origin-private-file-system/
- https://wicg.github.io/file-system-access/
- https://caniuse.com/native-filesystem-api
- https://github.com/sqlite/sqlite-wasm/blob/main/sqlite-wasm/jswasm/sqlite3-opfs-async-proxy.js

Recent discussions that can be solved by this API:

- #1433 
- #1387

**Comment by carlopi** [comment]
Thanks for raising this.

This would be very very interesting a getting this started is not too complex, but work has been mostly in other areas and this would need to be prioritised for one of the next release cycles.

**Comment by warrenbhw** [comment]
Any news on this? Would be very much interested in built-in local filesystem persistence.

**Comment by hangxingliu** [comment]
#1490
